spec/p2p/reactor/reactor.qnt

```diff
@@ -0,0 +1,238 @@
+/*
+ * Reactor is responsible for handling incoming messages on one or more
+ * Channel. Switch calls GetChannels when reactor is added to it. When a new
+ * peer joins our node, InitPeer and AddPeer are called. RemovePeer is called
+ * when the peer is stopped. Receive is called when a message is received on a
+ * channel associated with this reactor.
+ */
+// Code: https://github.com/cometbft/cometbft/blob/main/p2p/base_reactor.go
+module reactor {
+
+	// Unique ID of a node.
+	type NodeID = str
+
+	/*
+	 * Peer is an interface representing a peer connected on a reactor.
+	 */
+	type Peer = {
+		ID: NodeID,
+
+		// Other fields can be added to represent the p2p operation.
+	}
+
+	// Byte ID used by channels.
+	type Byte = int
+
+	// Channel configuration.
+	// Only the Byte ID is relevant here.
+	type ChannelDescriptor = {
+		ID: Byte,
+		Priority: int,
+	}
+
+	/*
+	 * Envelope contains a message with sender routing info.
+	 */
+	type Envelope = {
+		Src: Peer, // Sender
+		Message: str, // Payload
+		ChannelID: Byte,
+	}
+
+	// A Routine is used to interact with an active Peer.
+	type Routine = {
+		name: str,
+		peer: Peer,
+	}
+
+	type ReactorState = {
+		// Peers that have been initialized but not yet removed.
+		// The reactor should expect receiving messages from them.
+		peers: Set[NodeID],
+
+		// The reactor runs multiple routines.
+		routines: Set[Routine],
+
+		// Service implementation.
+		started: bool,
+		stopped: bool,
+
+		// Indicates whether a switch was set: reactor registration.
+		switch: bool,
+	}
+
+	var state: ReactorState
+
+	/*
+	 * GetChannels returns the list of MConnection.ChannelDescriptor. Make sure
+	 * that each ID is unique across all the reactors added to the switch.
+	 */
+	pure def GetChannels(s: ReactorState): Set[ChannelDescriptor] = {
+		// This method should return a list of channels, e.g.:
+		Set({ID: 3, Priority: 1})
```

Comment:
Thinking in terms of testing the Quint spec, instead of having this hardcoded example here, could have a set of ChanelDescriptor as part of the ReactorState and returned by this function.
The set would be initialized by the test functions.

Reply:
Yes, this is probably the best solution. Although one thing I don't like on it is that this value should be static, to never change.

Reply:
If it is part of the state, you can initialize it normally.
Another option is to have the channels as `const` which are initialized on `import`, like in the following example: 

```
module reactor {
   type test = int

   type State = {bla: test}

   const state: State
}

module tests {
    import reactor as R (
          state = {bla: 1}
    )
}
```

Mention it in case you want to use the feature in the future. Right now it is unstable and I don't recommend it.

Reply:
Is it even a possibility to change channel ids after the node is initialized? Is it possible that different nodes are configured with different ids for the same channel name? And that we need to model and analyze this scenario? If not, we don't need to model this possibility and channels in the spec should be fixed (`val`s) for all nodes.

Reply:
> Is it even a possibility to change channel ids after the node is initialized?

No. The Reactor expect a call to `GetChannels()` at the registration phase, which only happens once.

> channels in the spec should be fixed (vals) for all nodes.

Yes, this is a good solution.

Reply:
> Is it possible that different nodes are configured with different ids for the same channel name? 

It is possible but nothing will work in this case. Channels don't have name, to be precise, but the are assumed to be _global_, meaning that all nodes use the same channel byte ID to implement the same channels (e.g., the mempool channel).